src/sentry/sentry_metrics/indexer/cache.py

```diff
@@ -82,3 +95,77 @@ def delete(self, key: str, cache_namespace: str) -> None:
         self.cache.delete_many(cache_keys, version=self.version)
+
+
+class CachingIndexer(StringIndexer):
+    def __init__(self, cache: StringIndexerCache, indexer: StringIndexer) -> None:
+        self.cache = cache
+        self.indexer = indexer
+
+    def bulk_record(
+        self, use_case_id: UseCaseKey, org_strings: Mapping[int, Set[str]]
+    ) -> KeyResults:
+        cache_keys = KeyCollection(org_strings)
+        metrics.gauge("sentry_metrics.indexer.lookups_per_batch", value=cache_keys.size)
+        cache_key_strs = cache_keys.as_strings()
+        cache_results = self.cache.get_many(cache_key_strs, use_case_id.value)
+
+        hits = [k for k, v in cache_results.items() if v is not None]
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "true", "caller": "get_many_ids"},
+            amount=len(hits),
+        )
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "false", "caller": "get_many_ids"},
+            amount=len(cache_results) - len(hits),
+        )
+
+        # used to compare to pre org_id indexer cache fetch metric
+        metrics.incr(
+            _INDEXER_CACHE_FETCH_METRIC,
+            amount=cache_keys.size,
+        )
+
+        cache_key_results = KeyResults()
+        cache_key_results.add_key_results(
+            [KeyResult.from_string(k, v) for k, v in cache_results.items() if v is not None],
+            FetchType.CACHE_HIT,
+        )
+
+        db_record_keys = cache_key_results.get_unmapped_keys(cache_keys)
+
+        if db_record_keys.size == 0:
+            return cache_key_results
+
+        db_record_key_results = self.indexer.bulk_record(use_case_id, db_record_keys.mapping)
+        self.cache.set_many(
+            db_record_key_results.get_mapped_key_strings_to_ints(), use_case_id.value
+        )
+        return cache_key_results.merge(db_record_key_results)
+
+    def record(self, use_case_id: UseCaseKey, org_id: int, string: str) -> Optional[int]:
+        result = self.bulk_record(use_case_id=use_case_id, org_strings={org_id: {string}})
+        return result[org_id][string]
+
+    def resolve(self, use_case_id: UseCaseKey, org_id: int, string: str) -> Optional[int]:
+        key = f"{org_id}:{string}"
+        result = self.cache.get(key, use_case_id.value)
+
+        if result and isinstance(result, int):
+            metrics.incr(_INDEXER_CACHE_METRIC, tags={"cache_hit": "true", "caller": "resolve"})
+            return result
+
+        metrics.incr(_INDEXER_CACHE_METRIC, tags={"cache_hit": "false", "caller": "resolve"})
+        id = self.indexer.resolve(use_case_id, org_id, string)
+
+        if id is not None:
+            self.cache.set(key, id, use_case_id.value)
+
+        return id
+
+    def reverse_resolve(self, use_case_id: UseCaseKey, id: int) -> Optional[str]:
+        # XXX: We don't cache here at all -- never did in the original
+        # implementation either.
```

Comment:
The implication of this comment is a bit unclear to me. Are you suggesting we actually should be caching here? If so, let's make this a TODO. If not, no need to XXX (which I generally interpret to mean HACK or something that needs changing). I'd drop the comment about the original implementation too as it's confusing for those unfamiliar with the history of this.

Reply:
you're right. i was just coming across this oddity but i don't think it actually causes any problems. we don't have to fix it